lib/Parse/ParseDecl.cpp

```diff
@@ -3926,9 +3927,46 @@ Parser::parseDecl(ParseDeclOptions Flags,
         break;
       }
 
-      const bool IsProbablyFuncDecl =
-          Tok.isIdentifierOrUnderscore() || Tok.isAnyOperator();
+      bool IsProbablyFuncDecl = false;
+
+      if (Tok.isIdentifierOrUnderscore() || Tok.isAnyOperator()) {
+
+        // This is a probe, so backtrack when we're done
+        // and don't emit diagnostics.
+        BacktrackingScope backtrackingScope(*this);
+        DiagnosticSuppression diagnosticSuppression(Diags);
+
+        // TODO make sure we can't get these from the environment
+        SourceLoc staticLoc; 
+        StaticSpellingKind staticSpelling = StaticSpellingKind::None;
+        DeclAttributes attributes;
+        bool hasFuncKeyword = false;
+
+        ParserResult<FuncDecl> parserResult =
+          parseDeclFunc(staticLoc,
+            staticSpelling,
+            Flags,
+            attributes,
+            hasFuncKeyword);
+
+        // it looks like we always get a good parse result.
+        // But if there are no parentheses
+        // the left and right parens will be at the beginning of
+        // the expression and will be equal so we use
+        // this to reject degenerate declarations.
+
+        if (parserResult.isNonNull()) {
+          auto parameterList = parserResult.get()->getParameters();
+          if (parameterList->getLParenLoc() != parameterList->getRParenLoc()) {
+            IsProbablyFuncDecl = true;
+          }
+        }
```

Comment:
I don't quite follow the reasoning here. Why is the degenerate case rejected? Also, if the parser always succeeds, there must be something wrong with it... Not every syntactic form is a valid function declaration.

Reply:
Yeah, this is confusing. `parseDeclFunc` returns a non-null parse result even when it does not completely parse a function. It's pretty liberal in what it will accept. That's part of why this was happening originally--it thought the statement was a function then tried to parse it and sent out a lot of confusing diagnostics. 

The situation we are guarding against is a statement at the type level when there should be a function. `parseDeclFunc` is returning a non-null `ParserResult<FuncDecl>` in that situation and the only indicator (empirically) is that the left and right paren locations are identical in the `ParserResult<FuncDecl>`. 

I did not re-run all this through the debugger today. But as I remember it the parentheses are not the sensitive point--there are a couple of things that can happen in `parseFuncDecl` and still return a `ParserResult<FuncDecl>`. The location of the parens is just the differentiator the situation where we expect a function decl and can't parse one. You might also get a null `parserRequest` in other situations. 

Maybe `parseFuncDecl` should be tighter in what it accepts and return null when it does not get a full declaration. I'm not sure right how if that change would break anything else though. The code is not that intricate but the effects could be large. 

I'm sort of assuming that the normal clients of `parseFuncDecl` want to get as much information as possible out of it, while for this specific problem we want to inhibit some of those situations in the probe to skip to our diagnostic. 

Does that make sense?